test/sql-features/mzcompose.py

```diff
@@ -0,0 +1,225 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+import argparse
+import os
+import tempfile
+from textwrap import dedent, indent
+
+from materialize.mzcompose import Composition, WorkflowArgumentParser
+from materialize.mzcompose.services import Materialized, Testdrive
+
+SERVICES = [
+    Materialized(unsafe_mode=False),
+    Testdrive(no_reset=True, seed=1),
+]
+
+MZ_SYSTEM_CONNECTION_URL = (
+    "postgres://mz_system:materialize@${testdrive.materialize-internal-sql-addr}"
+)
+USER_CONNECTION_URL = (
+    "postgres://materialize:materialize@${testdrive.materialize-sql-addr}"
+)
+
+
+def header(test_name: str, drop_schema: bool) -> str:
+    """Generate a TD header for a SQL feature test scenario."""
+    header = dedent(
+        f"""
+        # Feature test for SQL feature test: {test_name}
+        #####################################{'#' * len(test_name)}
+        """
+    )
+    # Re-create schema (optional).
+    if drop_schema:
+        header += dedent(
+            f"""
+            > DROP SCHEMA IF EXISTS public CASCADE;
+            > CREATE SCHEMA public /* {test_name} */;
+            """
+        )
+    # Create connections.
+    header += dedent(
+        f"""
+        $ postgres-connect name=user url={USER_CONNECTION_URL}
+        $ postgres-connect name=mz_system url={MZ_SYSTEM_CONNECTION_URL}
+        """
+    )
+    return header.strip()
+
+
+def statement_error(statement: str, error_msg: str) -> str:
+    """Generte a TD command that asserts that `statement` fails with `error_msg`."""
+    return "\n".join(
+        [
+            f"! {indent(statement.strip(), prefix='  ')[2:]}",
+            f"contains:{error_msg}",
+        ]
+    )
+
+
+def statement_ok(statement: str) -> str:
+    """Generte a TD command that executes `statement`."""
+    return f"> {indent(statement.strip(), prefix='  ')[2:]}"
```

Comment:
It took me a while to figure out what is going on here and all the other places where a similar construct is used -- you are fishing for the second line returned by the `create_item()` method. It may be best to split this into two methods -- `create_item()` will return the `CREATE` portion and then an `invoke()` method can return the `SELECT` portion , which you can then use directly without fishing for it.

Reply:
I need to transform a possibly mutli-line SQL statement such as

```sql
CREATE TABLE t (
  x INT,
  y INT
)
```
to a valid TD command
```
> CREATE TABLE t (
    x INT,
    y INT
  )
```
which boils down to (1) indent the first line with '> ', and (2) indent all other lines with `  `. I'll see if this can be more readable.

Reply:
I think the caller needs to supply the statement fully-formatted so that you do not need to do that surgery.

Alternatively, you can run the statement using `$ postgres-execute` in which case everything up to the next `$` directive will be slurped and executed together.

Reply:
> I think the caller needs to supply the statement fully-formatted so that you do not need to do that surgery.

This violates the implicit contract of these helper functions: they receive data (SQL statements, expected errors, key-value pairs) and construct specific testdrive commands.

> Alternatively, you can run the statement using `$ postgres-execute` in which case everything up to the next $ directive will be slurped and executed together.

This is fine for the `statement_ok` function but not for the `statement_error` function. How about:

```python
indent(statement.strip(), prefix="  ").replace("  ", "> ", 1)
```
and
```python
indent(statement.strip(), prefix="  ").replace("  ", "! ", 1)
```

Reply:
Thanks, understood.